pkg/cloudprovider/providers/gce/gce_address_manager.go

```diff
@@ -0,0 +1,155 @@
+	// or the controller, but won't tell us the current state of the controller's IP. The address
+	// could be reserving another address; therefore, it would need to be deleted. In the normal
+	// case of using a controller address, retrieving the address by name results in the fewest API
+	// calls since it indicates whether a Delete is necessary before Reserve.
+	glog.V(3).Infof("%v: attempting hold of IP %v with type %v", am.logPrefix, am.targetIP, am.addressType)
+	// Get the address in case it was orphaned earlier
+	addr, err := am.svc.GetBetaRegionAddress(am.lbName, am.region)
+	if err != nil && !isNotFound(err) {
+		return err
+	}
+
+	if addr != nil {
+		// If address exists, check if the address IP is different from what's expected.
+		if addr.Address != am.targetIP || addr.AddressType != string(am.addressType) {
+			glog.V(3).Infof("%v: existing address %q has IP %v Type %v which does not match targeted IP %v Type %v. Attemping to delete.", am.logPrefix, addr.Name, addr.Address, addr.AddressType, am.targetIP, am.addressType)
+			err := am.svc.DeleteRegionAddress(addr.Name, am.region)
+			if err != nil {
+				if isNotFound(err) {
+					glog.V(3).Infof("%v: address %q was not found. Ignoring.", am.logPrefix, addr.Name)
+				} else {
+					return err
+				}
+			} else {
+				glog.V(3).Infof("%v: successfully deleted previous address %q", am.logPrefix, addr.Name)
+			}
+		} else {
+			glog.V(3).Infof("%v: address %q already reserves targeted IP %v of type %v. No further action required.", am.logPrefix, addr.Name, am.targetIP, am.addressType)
+			return nil
+		}
+	}
+
+	return am.ensureAddressReservation()
+}
+
+// ReleaseAddress will release the address if it's owned by the controller.
+func (am *addressManager) ReleaseAddress() error {
+	if !am.tryRelease {
+		glog.V(3).Infof("%v: not attempting release of address %v.", am.logPrefix, am.targetIP)
+		return nil
+	}
+
+	glog.V(3).Infof("%v: releasing address %v named %q", am.logPrefix, am.targetIP, am.lbName)
+	// Controller only ever tries to unreserve the address named with the load balancer's name.
+	err := am.svc.DeleteRegionAddress(am.lbName, am.region)
+	if err != nil {
+		if isNotFound(err) {
+			glog.Warningf("%v: address %q was not found. Ignoring.", am.logPrefix, am.targetIP, am.lbName)
+			return nil
+		}
+
+		return err
+	}
+
+	glog.V(3).Infof("%v: successfully released IP %v named %q", am.logPrefix, am.targetIP, am.lbName)
+	return nil
+}
+
+func (am *addressManager) ensureAddressReservation() error {
+	// Try reserving the IP with controller-owned address name
+	newAddr := &computebeta.Address{
+		Name:        am.lbName,
+		Description: fmt.Sprintf(`{"kubernetes.io/service-name":"%s"}`, am.serviceName),
+		Address:     am.targetIP,
+		AddressType: string(am.addressType),
+		Subnetwork:  am.subnetURL,
+	}
+
+	err := am.svc.ReserveBetaRegionAddress(newAddr, am.region)
+	if err != nil {
+		if !isHTTPErrorCode(err, http.StatusConflict) {
```

Comment:
Actually, this is not correct. I just found out that if you try to reserve an address with an IP that's already static (but not using the same name), you get a 400 error.
```
Error 400: Invalid value for field 'resource.address': '1.2.3.4'. Specified IP address is already reserved., invalid
```

Reply:
FWIW, I don't think the current external LB controller handle some corner cases because of this....
E.g., if the fwd rule is using an IP that's not specified in the Service object, but is actually owned by the user (e.g., not using the lbName), the controller would just fail to ensure the IP.

Reply:
Weird, I just tested this again and got 409. Wonder if there's an alpha vs beta difference
```
{
 "kind": "compute#operation",
 "id": "1599788789319664192",
 "name": "operation-1503706287098-5579ce716c890-3262ae10-677fa2b7",
 "operationType": "insert",
 "targetLink": "https://www.googleapis.com/compute/beta/projects/nicksardo-playground/regions/us-central1/addresses/myip2",
 "targetId": "5223319394054898240",
 "status": "DONE",
 "progress": 100,
 "insertTime": "2017-08-25T17:11:27.558-07:00",
 "startTime": "2017-08-25T17:11:28.371-07:00",
 "endTime": "2017-08-25T17:11:31.465-07:00",
 "error": {
  "errors": [
   {
    "code": "IP_IN_USE_BY_ANOTHER_RESOURCE",
    "message": "IP '10.0.0.10' is already being used by another resource. "
   }
  ]
 },
 "httpErrorStatusCode": 409,
 "httpErrorMessage": "CONFLICT",
 "selfLink": "https://www.googleapis.com/compute/beta/projects/nicksardo-playground/regions/us-central1/operations/operation-1503706287098-5579ce716c890-3262ae10-677fa2b7",
 "region": "https://www.googleapis.com/compute/beta/projects/nicksardo-playground/regions/us-central1"
}
```

Reply:
I tested more and it's not about alpha/beta/v1. The function returns 409 if AddressType == INTERNAL. Otherwise, it always returns 400. (sigh...)

Reply:
We should take this up with the appropriate team - it should be consistent.

Reply:
Handling both